### PR TITLE
Add bob.learn.em recipe

### PR DESCRIPTION
**Pull request by 183amir**

### CODE REVIEW

**Comment by 183amir** [comment]
This needs a careful review.

**Comment by jakirkham** [comment]
Seems it is building now so that fixed it then.

**Comment by 183amir** [comment]
This recipe builds fine on my laptop. Is the test being killed? Maybe it's taking too much time?
`staged-recipes (git)-[bob.learn.em] % conda build -c defaults -c conda-forge --override-channels --python=2.7 recipes/bob.learn.em` works fine.

**Comment by jakirkham** [comment]
So, this built all the way through, but you have test failures. It can't import `bob.learn.em`.

**Comment by 183amir** [comment]
I can't reproduce the bug which is weird.

**Comment by jakirkham** [comment]
So, as this is Linux, we have a docker container that you could use to play with this. It's on Docker Hub. Just pull down this image `pelson/obvious-ci:latest_x64`.

**Comment by 183amir** [comment]
Finally got the error message:
```
import bob.learn.em
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "bob/learn/em/__init__.py", line 9, in <module>
    bob.extension.load_bob_library('bob.learn.em', __file__)
  File "/opt/conda/envs/_test/lib/python2.7/site-packages/bob/extension/__init__.py", line 234, in load_bob_library
    ctypes.cdll.LoadLibrary(full_libname)
  File "/opt/conda/envs/_test/lib/python2.7/ctypes/__init__.py", line 443, in LoadLibrary
    return self._dlltype(name)
  File "/opt/conda/envs/_test/lib/python2.7/ctypes/__init__.py", line 365, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: bob/learn/em/libbob_learn_em.so: cannot open shared object file: No such file or directory
```
There is something wrong with the bash script `run_docker_build.sh` since this error was not being printed when I used it but I got this error message after I run commands by hand and exited the bash. (it was not showing me the error until I got out of bash.)
How can I run commands in docker and keep my modifications?

**Comment by 183amir** [comment]
Sorry that was not the error. The program is segfaulting `134 Segmentation fault`.

**Comment by 183amir** [comment]
Here is what I get when I try to run `import bob.learn.em`:
```
GNU gdb (GDB) Red Hat Enterprise Linux (7.2-83.el6)
Copyright (C) 2010 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>...
Reading symbols from /opt/conda/envs/_test/bin/python...done.
Starting program: /opt/conda/envs/_test/bin/python /test.py
warning: no loadable sections found in added symbol-file system-supplied DSO at 0x7ffff7ffa000
[Thread debugging using libthread_db enabled]
Detaching after fork from child process 174.
Missing separate debuginfo for /opt/conda/envs/_test/lib/python2.7/site-packages/numpy/core/../../../.././libgfortran.so.3
[New Thread 0x7ffff0176700 (LWP 176)]
[New Thread 0x7fffef975700 (LWP 177)]
[New Thread 0x7fffed174700 (LWP 178)]
580	/opt/conda/envs/_build/lib/python2.7/site-packages/bob/extension/include/bob.extension/documentation.h: No such file or directory.

Program received signal SIGSEGV, Segmentation fault.
bob::extension::FunctionDoc::doc(unsigned int, unsigned int) const () at /opt/conda/envs/_build/lib/python2.7/site-packages/bob/extension/include/bob.extension/documentation.h:580
	in /opt/conda/envs/_build/lib/python2.7/site-packages/bob/extension/include/bob.extension/documentation.h
Missing separate debuginfos, use: debuginfo-install glibc-2.12-1.166.el6_7.7.x86_64
(gdb) bt
#0  bob::extension::FunctionDoc::doc(unsigned int, unsigned int) const () at /opt/conda/envs/_build/lib/python2.7/site-packages/bob/extension/include/bob.extension/documentation.h:580
#1  0x00007fffe441477b in _GLOBAL__sub_I_main.cpp () at bob/learn/em/main.cpp:18
#2  0x00007fffe4414826 in __do_global_ctors_aux () from /opt/conda/envs/_test/lib/python2.7/site-packages/bob/learn/em/_library.so
#3  0x00007fffe4391d03 in _init () from /opt/conda/envs/_test/lib/python2.7/site-packages/bob/learn/em/_library.so
#4  0x00007fff00000000 in ?? ()
#5  0x00007ffff7deb625 in _dl_init_internal () from /lib64/ld-linux-x86-64.so.2
#6  0x00007ffff7defe95 in dl_open_worker () from /lib64/ld-linux-x86-64.so.2
#7  0x00007ffff7deb286 in _dl_catch_error () from /lib64/ld-linux-x86-64.so.2
#8  0x00007ffff7def63a in _dl_open () from /lib64/ld-linux-x86-64.so.2
#9  0x00007ffff75d5f66 in dlopen_doit () from /lib64/libdl.so.2
#10 0x00007ffff7deb286 in _dl_catch_error () from /lib64/ld-linux-x86-64.so.2
#11 0x00007ffff75d629c in _dlerror_run () from /lib64/libdl.so.2
#12 0x00007ffff75d5ee1 in dlopen@@GLIBC_2.2.5 () from /lib64/libdl.so.2
#13 0x00007ffff7b25dde in _PyImport_GetDynLoadFunc (fqname=<value optimized out>, shortname=<value optimized out>, pathname=0xd34d80 "/opt/conda/envs/_test/lib/python2.7/site-packages/bob/learn/em/_library.so", fp=0xd39850) at Python/dynload_shlib.c:130
#14 0x00007ffff7b0a8d8 in _PyImport_LoadDynamicModule (name=0xcfa0d0 "bob.learn.em._library", pathname=0xd34d80 "/opt/conda/envs/_test/lib/python2.7/site-packages/bob/learn/em/_library.so", fp=0xd39850) at ./Python/importdl.c:42
#15 0x00007ffff7b08f81 in import_submodule (mod=0x7ffff7eaada8, subname=0xcfa0dd "_library", fullname=0xcfa0d0 "bob.learn.em._library") at Python/import.c:2704
#16 0x00007ffff7b091f4 in load_next (mod=0x7ffff7eaada8, altmod=0x7ffff7eaada8, p_name=<value optimized out>, buf=0xcfa0d0 "bob.learn.em._library", p_buflen=0x7fffffffe0b0) at Python/import.c:2519
#17 0x00007ffff7b09820 in import_module_level (name=<value optimized out>, globals=<value optimized out>, locals=<value optimized out>, fromlist=0x7ffff7ef6a10, level=<value optimized out>) at Python/import.c:2228
#18 PyImport_ImportModuleLevel (name=<value optimized out>, globals=<value optimized out>, locals=<value optimized out>, fromlist=0x7ffff7ef6a10, level=<value optimized out>) at Python/import.c:2292
#19 0x00007ffff7ae914f in builtin___import__ (self=<value optimized out>, args=<value optimized out>, kwds=<value optimized out>) at Python/bltinmodule.c:49
#20 0x00007ffff7a3fd23 in PyObject_Call (func=0x7ffff7fdffc8, arg=<value optimized out>, kw=<value optimized out>) at Objects/abstract.c:2546
#21 0x00007ffff7ae9633 in PyEval_CallObjectWithKeywords (func=0x7ffff7fdffc8, arg=0x7ffff7e935f0, kw=<value optimized out>) at Python/ceval.c:4219
#22 0x00007ffff7aee29e in PyEval_EvalFrameEx (f=<value optimized out>, throwflag=<value optimized out>) at Python/ceval.c:2622
#23 0x00007ffff7af3a2e in PyEval_EvalCodeEx (co=0x7ffff7eb0930, globals=<value optimized out>, locals=<value optimized out>, args=<value optimized out>, argcount=0, kws=0x0, kwcount=0, defs=0x0, defcount=0, closure=0x0) at Python/ceval.c:3582
#24 0x00007ffff7af3b42 in PyEval_EvalCode (co=<value optimized out>, globals=<value optimized out>, locals=<value optimized out>) at Python/ceval.c:669
#25 0x00007ffff7b05a82 in PyImport_ExecCodeModuleEx (name=0x63f160 "bob.learn.em", co=0x7ffff7eb0930, pathname=0x6b9400 "/opt/conda/envs/_test/lib/python2.7/site-packages/bob/learn/em/__init__.pyc") at Python/import.c:713
#26 0x00007ffff7b081ce in load_source_module (name=0x63f160 "bob.learn.em", pathname=0x6b9400 "/opt/conda/envs/_test/lib/python2.7/site-packages/bob/learn/em/__init__.pyc", fp=<value optimized out>) at Python/import.c:1103
#27 0x00007ffff7b08a2a in load_package (name=0x63f160 "bob.learn.em", pathname=<value optimized out>) at Python/import.c:1170
#28 0x00007ffff7b08f81 in import_submodule (mod=0x7ffff7eaad70, subname=0x63f16a "em", fullname=0x63f160 "bob.learn.em") at Python/import.c:2704
#29 0x00007ffff7b091f4 in load_next (mod=0x7ffff7eaad70, altmod=0x7ffff7eaad70, p_name=<value optimized out>, buf=0x63f160 "bob.learn.em", p_buflen=0x7fffffffe6a0) at Python/import.c:2519
#30 0x00007ffff7b09860 in import_module_level (name=<value optimized out>, globals=<value optimized out>, locals=<value optimized out>, fromlist=0x7ffff7da6cd0, level=<value optimized out>) at Python/import.c:2236
#31 PyImport_ImportModuleLevel (name=<value optimized out>, globals=<value optimized out>, locals=<value optimized out>, fromlist=0x7ffff7da6cd0, level=<value optimized out>) at Python/import.c:2292
#32 0x00007ffff7ae914f in builtin___import__ (self=<value optimized out>, args=<value optimized out>, kwds=<value optimized out>) at Python/bltinmodule.c:49
#33 0x00007ffff7a3fd23 in PyObject_Call (func=0x7ffff7fdffc8, arg=<value optimized out>, kw=<value optimized out>) at Objects/abstract.c:2546
#34 0x00007ffff7ae9633 in PyEval_CallObjectWithKeywords (func=0x7ffff7fdffc8, arg=0x7ffff7ed15d0, kw=<value optimized out>) at Python/ceval.c:4219
#35 0x00007ffff7aee29e in PyEval_EvalFrameEx (f=<value optimized out>, throwflag=<value optimized out>) at Python/ceval.c:2622
#36 0x00007ffff7af3a2e in PyEval_EvalCodeEx (co=0x7ffff7edceb0, globals=<value optimized out>, locals=<value optimized out>, args=<value optimized out>, argcount=0, kws=0x0, kwcount=0, defs=0x0, defcount=0, closure=0x0) at Python/ceval.c:3582
#37 0x00007ffff7af3b42 in PyEval_EvalCode (co=<value optimized out>, globals=<value optimized out>, locals=<value optimized out>) at Python/ceval.c:669
#38 0x00007ffff7b14050 in run_mod (fp=0x61f240, filename=<value optimized out>, start=<value optimized out>, globals=0x7ffff7f81168, locals=0x7ffff7f81168, closeit=1, flags=0x7fffffffebc0) at Python/pythonrun.c:1370
#39 PyRun_FileExFlags (fp=0x61f240, filename=<value optimized out>, start=<value optimized out>, globals=0x7ffff7f81168, locals=0x7ffff7f81168, closeit=1, flags=0x7fffffffebc0) at Python/pythonrun.c:1356
#40 0x00007ffff7b1422f in PyRun_SimpleFileExFlags (fp=0x61f240, filename=0x7fffffffeed2 "/test.py", closeit=1, flags=0x7fffffffebc0) at Python/pythonrun.c:948
#41 0x00007ffff7b29b74 in Py_Main (argc=<value optimized out>, argv=<value optimized out>) at Modules/main.c:645
#42 0x00007ffff6dd8d5d in __libc_start_main () from /lib64/libc.so.6
#43 0x0000000000400649 in _start ()
```

**Comment by 183amir** [comment]
Looks good?

**Comment by 183amir** [comment]
Awesome the new gcc fixed this pull too.

**Comment by jakirkham** [comment]
I was going to ask about the patch, but I guess that is already gone. Glad to see this builds cleanly without it.

I think I am ok with this, but would like to fix the rest of the existing bob feedstocks (merge the `gcc` PRs) first and figure out one other point. Is this linking to a BLAS?

**Comment by 183amir** [comment]
I'm done with removing gcc from bob packages.
How would I check if this is linking against blas? I think it is using bob.math api where that one uses blas.

**Comment by 183amir** [comment]
This does not link against blas. This is linking arguments: `-lbob_learn_em -lbob_learn_activation -lbob_sp -lbob_math -lbob_learn_linear -lbob_core -lbob_io_base -l:libboost_system.so.1.60.0 -lblitz -lpython2.7 -o build/lib.linux-x86_64-2.7/bob/learn/em/_library.so`

File: recipes/bob.learn.em/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    - bob.learn
+    - bob.learn.em
+    - bob.learn.em.test
+
+  requires:
+    - nose
+
+about:
+  home: http://github.com/bioidiap/bob.learn.em
+  license: Modified BSD License (3-clause)
```
Comment (msarahan) [inline recipes/bob.learn.em/meta.yaml]: Shorten license to 3-clause BSD (if you want.  If not, say so and leave it)

**Comment by msarahan** [comment]
Another good one here.  +1 for merge - license shortening question excepted.

File: recipes/bob.learn.em/meta.yaml (hunk 1 above)
Comment (jakirkham) [inline recipes/bob.learn.em/meta.yaml]: Same [story]( https://github.com/conda-forge/staged-recipes/pull/494#discussion_r61833638 ).

**Comment by jakirkham** [comment]
Thanks @183amir.

**Pull request merged**